Add docker image

davideicardi opened the pull request:
Fix #191

davideicardi commented:
I have tested it locally but not sure if github action script is correct ... Not sure how to test it.

davideicardi commented:
And you have to set `DOCKER_TOKEN` secret I suppose.

Anthony-Mckale commented:
really nice guessing once this is approved owner can create

https://hub.docker.com/r/cyclonedx/cyclonedx-node

to match https://hub.docker.com/r/cyclonedx/cyclonedx-python

Anthony-Mckale commented:
would perhaps add a "docker-release" command to the scripts but that's just me being lazy for the project i maintain

something like

```json
"scripts": {
...
"docker-release": "docker build . -t cyclonedx/cyclonedx-node:`node bin/cyclonedx-bom --version` -t cyclonedx/cyclonedx-node:latest && docker push riversafe/eze-cli:`node bin/cyclonedx-bom --version` && docker push cyclonedx/cyclonedx-node:latest"
}
```

davideicardi commented:
Thank you @Anthony-Mckale. Just added `docker-release` script as suggested.

davideicardi commented:
News? It is possible to publish the official docker image? Thank you!

jkowalleck commented:
@coderpatros @stevespringett your attention is needed

stevespringett commented on .github/workflows/release.yml:
```diff
@@ -46,13 +46,27 @@ jobs:
+      run: |
+        REPO=cyclonedx/cyclonedx-node
+        VERSION=${{ steps.build.outputs.version }}
+        docker login --username coderpatros --password "$DOCKER_TOKEN"
```
`coderpatros` needs to be changed to `$DOCKERHUB_USERNAME `
`$DOCKER_TOKEN` needs to be changed to `$DOCKERHUB_TOKEN`.

Both DOCKERHUB_USERNAME and DOCKERHUB_TOKEN are secrets defined at the GitHub organization level. I assume DOCKER_TOKEN is likely unique to this repo and should be deleted in favor of the organizational secrets.

davideicardi replied:
Done! I think that for now this repo doesn't yet publish to docker so no variables was defined.

stevespringett commented on .github/workflows/release.yml:
```diff
@@ -46,13 +46,27 @@ jobs:
+
+    - name: Publish Docker image to Docker Hub
+      env:
+        DOCKER_TOKEN: ${{ secrets.DOCKER_TOKEN }}
```
`DOCKER_TOKEN` needs to be changed to `secrets.DOCKERHUB_TOKEN`.
And a new `DOCKERHUB_USERNAME` needs to be created and set to `secrets. DOCKERHUB_USERNAME `.

davideicardi replied:
Done!

jkowalleck commented on README.md:
````diff
@@ -57,6 +57,20 @@ cyclonedx-bom -o bom.xml
+
+```bash
+docker run --rm \
+  -v `pwd`:/src \
````
add `""` around `pwd` - to have spaces in paths escaped
better go with `"$PWD"` ? it is universal, and unlike `pwd` it works in shells that are not bash.

davideicardi replied:
Yes, I think you are right. I will fix it.

jkowalleck commented on README.md:
```diff
@@ -57,6 +57,20 @@ cyclonedx-bom -o bom.xml
+docker run --rm \
+  -v `pwd`:/src \
+  -w /src \
+  cyclonedx/cyclonedx-node -o /src/bom.xml /src
```
i'd recomment to remove `/src` from the arguments here, since `/src` is already set as the working-dir (`-w /src`)

```diff
- cyclonedx/cyclonedx-node -o /src/bom.xml /src
+ cyclonedx/cyclonedx-node -o bom.xml
```

davideicardi replied:
Yes, good point.

jkowalleck commented:
thanks again, @davideicardi 
your effort is appreciated.

I've added some remarks in the files, that you need to consider. 
Additionally, could you also add a docker watcher for dependabot to `.github/dependabot.yml` ?

davideicardi commented on .github/workflows/release.yml:
```diff
@@ -46,13 +46,28 @@ jobs:
         npm run build --if-present
-        npm version ${{ github.event.inputs.versionLevel }} --message "${{ github.event.inputs.commitMessage }}"
+        VERSION=$(npm version ${{ github.event.inputs.versionLevel }} --message "${{ github.event.inputs.commitMessage }}")
+        VERSION=${VERSION:1} # remove 'v' prefix
```
I think it is just a convention, usually docker tags doesn't have the `v` prefix. Also other CycloneDx images follow the same rules (https://hub.docker.com/r/cyclonedx/cyclonedx-python/tags, https://hub.docker.com/r/cyclonedx/cyclonedx-dotnet/tags, ...). Personally I prefer to remove the `v`.

But I understand your point so we can do whatever you prefer ;-).

Please let me know if I should leave the `v` prefix or not.

davideicardi commented on Dockerfile:
```diff
@@ -0,0 +1,20 @@
+COPY . /usr/src/cyclonedx-bom
+
+ENTRYPOINT ["/usr/src/cyclonedx-bom/bin/cyclonedx-bom"]
+CMD ["-h"]
```
Yes, right now if you do not add any arguments it will just show  the help.

In my opinion is a good practice to avoid unexpected behavior and it is the same logic used by [cyclonedx-dotnet](https://github.com/CycloneDX/cyclonedx-dotnet/blob/master/Dockerfile).

But also in this case it is not a strong opinion, as you prefer.

davideicardi commented:
@jkowalleck Thank you for your feedbacks.

I have added dependabot and fixed or answered to your points.

Happy to contribute to cyclonedx ecosystem!

stevespringett commented:
Big thanks for this PR and for all the feedback. Much appreciated.
